aip/general/0202.md

```diff
@@ -0,0 +1,103 @@
+---
+id: 202
+state: draft
+created: 2023-09-14
+placement:
+  category: fields
+  order: 0
+---
+
+# Fields
+
+The [`google.api.Field`][field proto] type, through its accompanying extension
+`google.api.field`, enriches a field's schema beyond the basic name and type
+information.
+
+## Guidance
+
+Decorating a field with `google.api.field` is only necessary when explicitly
+stated in this AIP or another that leverages `google.api.Field` information. As
+such, the guidance herein applies to those scenarios as well.
+
+### Format
+
+Fields with a primitive type can still have a specific format. To convey that
+type format, the `Field.Format` enumeration is used via the
+`(google.api.field).format` extension field. The following guidance conveys the
+meaning of and requirements for use of each `Field.Format` value.
+
+#### UUID4
+
+The `UUID4` format represents a UUID version 4 value as governed by
+[RFC 4122][]. It **must** only be used on a field of type `string`.
+
```

Comment:
TL;DR --- equivalency requirement is more important to establish than normalization because **equivalency** as a hard requirement permits **normalization** as a consequence ---- but **normalization** alone being permissible is still breaking to processors which have not put the **equivalency** consideration in their payload comparison logic.

Reply:
Do we feel comfortable publishing (or pointing at) equivalency functions for each of these types and all types we want to add in the future? Some of the hesitancy around stating equivalency on my end has been that we may not be able to compute equivalency on the client side for these.

For example, email canonicalization is an unsolvable problem from the client side

Reply:
I think including equivalence language is OK if we stay very far from the details of it i.e. the RFC states how equivalence is determined and all we say is "Consult RFC XXXX for equivalence semantics". But I agree with @slevenick in that it is a very slippery slope, one we cannot be expected to climb, if we include anything more specific than that.

The reason I started by only speaking to normalization is that it stays out of the nitty gritty of semantic details that could impact equivalence checks. For example, I don't want to include stuff like "`localhost` is a valid alias for the IPv4 address 127.0.0.1". The "primitive normalization" included, at the very least, enables the basic equivalence that Sam has mentioned before. Anything more than that should be done by dedicated, RFC-compliant libraries e.g. UUID parses which this AIP should not speak to imo. Showing consumers what it is gets them most of the way there. Having semi-uniform/normalized values gets them slightly further.

A general point about the `**may**` - we don't have infrastructure to enforce/enable this ATM and we have no idea what services are doing currently, so this is very conservative language that points folks towards where we want to go.

Reply:
> For example, email canonicalization is an unsolvable problem from the client side

Canonicalization means that that we attempting to define equivalence in terms of normalization. You are correct, that is impossible.

`true == equivalent<string>(normalize<T>(a), normalize<T>(b))`

That's exactly why I'm asserting we *do not* establish normalization as the basis.

Instead, establish equivalence as the basis, which *can* be defined from the client side. Normalization is then defined by equivalence, not the other way around.

`true == equals<T>(parse<T>(a), parse<T>(b))`

the definition of equivalent now determines what normalization is permitted. i.e. any change to the value which remains equivalent is allowed.

`b := normalize<T>(a).toString()`
`true == equals<T>(parse<T>(a), parse<T>(b))`

> The reason I started by only speaking to normalization is that it stays out of the nitty gritty of semantic details that could impact equivalence checks.  ... The "primitive normalization" included, at the very least, enables the basic equivalence that Sam has mentioned before.

I get the intent --- but it's having the opposite effect.

Reply:
Ok after some good offline discussion I will be adding language regarding equality that is specific in that it refers to the RFC for the topic, but vague enough that it doesn't provide an implementation for consumers (this is not the business of this AIP).

The relationship of "Services may normalize in ABC way, so consumers should compare values according to RFC 123" is the what we want to highlight.

Reply:
Ok, so...some RFCs had detailed language about format, others had language about comparisons...I decided to instead *discourage* primitive text comparison, and *encourage* leveraging a RFC compliant implementation for parsing.